pyacq/core/stream.py

```diff
@@ -505,29 +572,400 @@ def __init__(self, socket, params):
+        the step is negative. This makes it possible to collect the result in
+        the forward direction and handle the step later.
+        """
+        start_index = self._write_index + 1 - self.shape[0]
+        if isinstance(index, int):
+            if index < 0:
+                index += self._read_index + 1
+            if index > self._read_index or index < start_index:
+                raise IndexError("Index %d is out of bounds for ring buffer [%d, %d]" %
+                                 (index, start_index, self._read_index))
+            return index
+        elif isinstance(index, slice):
+            start, stop, step = index.start, index.stop, index.step
+            
+            # Handle None and negative steps
+            if step is None:
+                step = 1
+            if step < 0:
+                start, stop = stop, start
+                
+            # Interpret None and negative indices
+            if start is None:
+                start = start_index
+            else:
+                if start < 0:
+                    start += self._read_index + 1
+                if step < 0:
+                    start += 1 
+                
+            if stop is None:
+                stop = self._read_index + 1
+            else:
+                if stop < 0:
+                    stop += self._read_index + 1
+                if step < 0:
+                    stop += 1
+                
+            # Bounds check.
+            # Perhaps we could clip the returned data like lists/arrays do,
+            # but in this case the feedback is likely to be useful to the user.
+            if stop > self._read_index+1 or stop < start_index:
+                raise IndexError("Stop index %d is out of bounds for ring buffer [%d, %d]" %
+                                 (stop, start_index, self._read_index))
+            if start > self._read_index or start < start_index:
+                raise IndexError("Start index %d is out of bounds for ring buffer [%d, %d]" %
+                                 (start, start_index, self._read_index))
+            return start, stop, step
+        else:
+            raise TypeError("Invalid index %s" % index)
+        
+    def get_data(self, size, copy=False):
+        """Return *size* most recent data frames.
+        
+        If *copy* is True, then the returned array is copied from the ring buffer
+        and is safe to be modified. Otherwise, the returned buffer may be a
+        direct view on the internal ring buffer.
+        """
+        assert size <= self.size, ("Requested data size (%d) is larger than ring buffer (%d)." %
+                                   size, self.size)
+        if self.double:
+            ind = (self._ptr - size) % self.size
+            data = self.buffer[ind:ind+size]
+            if copy:
+                data = data.copy()
+        else:
+            ind = self._ptr - size
```

Comment:
Proposal: in the case double=False and ind<0 to avoid copy.
We could add an optional always_split=False.
get_data(self, size, copy=False, always_split=False):

always_split=False: is the case implemnted here
always_split=True return always 2 parts :
data1, data2 = RingBuffer.get_data(...)
where data2 is None except when ind<0.